### PR TITLE
Feature/DRM rest endpoints

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds two REST endpoints for DRM - one for listing terms and the other one for accepting terms.

Screenshot from 2021-07-12 15-21-20
Screenshot from 2021-07-12 15-21-45
Screenshot from 2021-07-12 15-22-50
Screenshot from 2021-07-12 15-23-12
Screenshot from 2021-07-12 15-23-29
Screenshot from 2021-07-12 15-24-04

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  @POST
+  @Path("/{uuid}/{version}")
+  def acceptDRM(@ApiParam("Item UUID") @PathParam("uuid") uuid: String,
+                @ApiParam("Item Version") @PathParam("version") version: Int): Response = {
+    try {
+      val item = LegacyGuice.itemService.getUnsecure(new ItemId(uuid, version))
+      Response.ok().entity(drmService.acceptLicenseOrThrow(item)).build()
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala:64]: I guess it is enough to just return the ID of new `DrmAcceptance` instance ?

**Comment by PenghaiZhang** [comment]
I have updated a new screenshot for the listing endpoint.

image

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    } catch {
+      case e: BadRequestException   => badRequest(e.getMessage)
+      case e: AccessDeniedException => forbiddenRequest(e.getMessage)
+      case e: DRMException          => unauthorizedRequest(e.getMessage)
+      case e: ItemNotFoundException => resourceNotFound(e.getMessage)
+    }
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala:70]: The rather repetitive catch block caught my eye, and it got me thinking, "could we use scala's FP capabilities to improve this?"

So how about something roughly like follows.

First, for the class define a HOF to map the exceptions:

```scala
def mapException(e: Exception) => (errors: String*): Response => e match {
  case BadRequestException => badRequest
  case AccessDeniedException => forbiddenRequest
  case DRMException => unauthorizedRequest
  case ItemNotFoundException => resourceNotFound
}
```

(If order not important though, please sort those - I just copied your order.)

```scala
allCatch.withTry {
  (LegacyGuice.itemService.getUnsecure andThen drmService.acceptLicenseOrThrow)(ItemId(uuid, version))
} match {
  case Failure(e) => mapException(e)(e.getMessage)
  case Success(item) => Response.ok().entity(item).build()
}
```

And if needed, you could replace `allCatch` with a custom catcher to just limit to those exceptions in `mapException`. See more @ https://www.baeldung.com/scala/exception-handling

Also, doing this, you can then use the same exception handling in the other DRM endpoints for consistent error handling.

(There's a few bits that could be further improved here - it'd be great to thread the value through, and repeated references to`e` in the `Failure(e)` clause - and there might be some syntax issues as I just roughed it up. But hopefully you can see what I'm aiming for.)

**Review by edalex-ian — CHANGES_REQUESTED** [review]
The content of this PR is good, but a few key considerations:

1. Lots of missing Java/ScalaDoc;
2. Missing API documentation - especially at the verb level;
3. This should probably be an endpoint under `/item/{uuid}/{version}/drm`. It doesn't have to be nestled into the 'Item' API though, it could possibly still be stand-alone and be 'Item DRM' just like there's 'Item Actions'.
4. I note in the Swagger API docs there's already mention of a number of DRM related models/beans - such as DRMOptions. Did you consider using some of them?
5. The one small code mentioned to refactor `acceptDRM` to aim for more functional code.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala (hunk 2 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/drm/DrmResource.scala]: Thanks Ian for this suggestion!! I have use `Try` together with some function compositions to improve !

**Review by edalex-ian — APPROVED** [review]
Good stuff @PenghaiZhang :+1: :tada:

**Pull request merged**